src/HypreLinearSystem.C

```diff
@@ -140,83 +126,543 @@ HypreLinearSystem::beginLinearSystemConstruction()
+				     std::vector<HypreIntType> & columns) {
+  for (unsigned i=0; i<numNodes; ++i) {
+    hids[i] = get_entity_hypre_id(nodes[i]);
+    for (unsigned d=0; d<numDof_; ++d)
+      columns[i*numDof_+d] = hids[i]*numDof_+d;
+  }
+}
+
 void
 HypreLinearSystem::buildNodeGraph(
-  const stk::mesh::PartVector&)
+  const stk::mesh::PartVector & parts)
 {
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  /* record the start time */
+  struct timeval _start, _stop;
+  gettimeofday(&_start, NULL);
+#endif
+
   beginLinearSystemConstruction();
+  stk::mesh::MetaData & metaData = realm_.meta_data();
+  const stk::mesh::Selector s_owned = metaData.locally_owned_part()
+    & stk::mesh::selectUnion(parts)
+    & !(stk::mesh::selectUnion(realm_.get_slave_part_vector()))
+    & !(realm_.get_inactive_selector());
+
+  stk::mesh::BucketVector const& buckets =
+    realm_.get_buckets( stk::topology::NODE_RANK, s_owned );
+
+  if (numDof_==1) {
+    std::vector<HypreIntType> hids(1);
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+	stk::mesh::Entity node = b[k];
+	hids[0] = get_entity_hypre_id(node);
+	
+	/* fill owned/shared 1 Dof */
+	fill_owned_shared_data_structures_1DoF(1, hids);
+      }
+    }
+  } else {
+    std::vector<HypreIntType> hids(1);
+    std::vector<HypreIntType> columns(numDof_);
+
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+	stk::mesh::Entity node = b[k];
+	hids[0] = get_entity_hypre_id(node);
+	for (unsigned d=0; d<numDof_; ++d) columns[d] = hids[0]*numDof_+d;
+
+	  /* fill owned/shared for more than 1 Dof */
+	fill_owned_shared_data_structures(1, hids, columns);
+      }
+    }
+  }
+
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  gettimeofday(&_stop, NULL);
+  double msec = (double)(_stop.tv_usec - _start.tv_usec) / 1.e3 + 1.e3*((double)(_stop.tv_sec - _start.tv_sec));
+  printf("rank_=%d EqnName=%s : %s %s %d : dt=%1.5lf\n",rank_,name_.c_str(),__FILE__,__FUNCTION__,__LINE__,msec);
+#endif
 }
 
+
 void
-HypreLinearSystem::buildFaceToNodeGraph(
-  const stk::mesh::PartVector&)
+HypreLinearSystem::buildFaceToNodeGraph(const stk::mesh::PartVector & parts)
 {
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  /* record the start time */
+  struct timeval _start, _stop;
+  gettimeofday(&_start, NULL);
+#endif
+
   beginLinearSystemConstruction();
+  stk::mesh::MetaData & metaData = realm_.meta_data();
+  const stk::mesh::Selector s_owned = metaData.locally_owned_part()
+                                      & stk::mesh::selectUnion(parts)
+                                      & !(realm_.get_inactive_selector());
+  stk::mesh::BucketVector const& buckets = realm_.get_buckets(realm_.meta_data().side_rank(), s_owned);
+
+  if (numDof_==1) {
+    std::vector<HypreIntType> hids(0);
+    if (buckets.size()) {
```

Comment:
The sizing of this 'hids' array needs to be inside the outer loop over buckets (at line 275). The buckets contain sides, which can be either quads or tris if the mesh has a mixture of element types. Within each bucket the sides are all the same, but different buckets can have different topologies. So after line 275 you can say 'numNodes = b.topology().num_nodes();' and do the hids.resize there, and then hids will always be the correct size in the inner loop at line 284.

Reply:
I sized the arrays outside of the big loops due to speed concerns. I hadn't encountered situations where the node count is different for a particular build*Graph algorithm?

Should I go in and fix these issues in my branch?

Reply:
I usually rank correctness above speed. If you put the resize inside the outer loop it's not much of a speed hit. If all your cases have homogeneous sides then it's no speed hit at all because a vector resize is very cheap if the size is staying the same.

Reply:
Agreed. I think I assumed that for methods like buildEdgeToNodeGraph, each element in the bucket would have the same number of nodes.

Reply:
fixed